cpp/src/arrow/ipc/reader.cc

```diff
@@ -717,8 +717,10 @@ Status ReadDictionary(const Buffer& metadata, DictionaryMemo* dictionary_memo,
     return Status::Invalid("Dictionary record batch must only contain one field");
   }
   auto dictionary = batch->column(0);
-  // Validate the dictionary for safe delta concatenation
-  RETURN_NOT_OK(dictionary->Validate());
+  // Validate the dictionary for safe delta concatenation.  Full validation
+  // is necessary for certain types (for example nested dictionaries).
+  // XXX: instead only validate when a concatenation is actually done?
+  RETURN_NOT_OK(dictionary->ValidateFull());
```

Comment:
Could this be deferred until the actual time that the delta must be applied? That seems better but I'm OK with this if a stopgap is needed

Reply:
Ok, it is the strategy I have finally implemented.